DOC: Fixed type hints for GroupBy.{any, all}

ytakashina opened the pull request:
The current documentation says the return value of GroupBy.all and GroupBy.any is `bool`.

```
    def any(self, skipna: bool = True):
        """
        Return True if any value in the group is truthful, else False.

        Parameters
        ----------
        skipna : bool, default True
            Flag to ignore nan values during truth testing.

        Returns
        -------
        bool
        """
```

However, the actual returned type is DataFrame or Series (with the same shape as `GroupBy.sum()`).

```
In [25]: df
Out[25]:
     0    1    2    3
0  1.0  0.0  0.0  0.0
1  0.0  1.0  0.0  0.0
2  0.0  0.0  1.0  0.0
3  0.0  0.0  0.0  1.0

In [26]: df.groupby(0).any()
Out[26]:
         1      2      3
0
0.0   True   True   True
1.0  False  False  False

In [27]: type(df.groupby(0).any())
Out[27]: pandas.core.frame.DataFrame
```

This PR fixes the wrong type hints for `GroupBy.{any, all}`.

- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

ytakashina committed:
DOC: Fixed type hints for GroupBy.{any, all}

The current documentation says the return value of GroupBy.all and GroupBy.any is `bool`.

```
    def any(self, skipna: bool = True):
        """
        Return True if any value in the group is truthful, else False.

        Parameters
        ----------
        skipna : bool, default True
            Flag to ignore nan values during truth testing.

        Returns
        -------
        bool
        """
```

However, the actual returned type is DataFrame or Series (the same shape as `GroupBy.sum()`).

```
In [25]: df
Out[25]:
     0    1    2    3
0  1.0  0.0  0.0  0.0
1  0.0  1.0  0.0  0.0
2  0.0  0.0  1.0  0.0
3  0.0  0.0  0.0  1.0

In [26]: df.groupby(0).any()
Out[26]:
         1      2      3
0
0.0   True   True   True
1.0  False  False  False

In [27]: type(df.groupby(0).any())
Out[27]: pandas.core.frame.DataFrame
```

This PR fixes the wrong type hints for `GroupBy.{any, all}`.

phofl commented:
That is a bit short imho. Could you maybe write something like here ? https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.GroupBy.sum.html

ytakashina commented:
Okay. I wrote some drafts for `GroupBy.all()` documentation. Which one do you think better?

1. Computed all of values within each group.
    - just replaced `sum` by `all` in GroupBy.sum documentation.
2. Computed all() of values within each group.
    - added parentheses to make it clear that `all` represents somewhat function (that is, it's not a pronoun).
3. DataFrame or Series of boolean values, which is True if all elements are True within each group, False otherwise.
    - wrote whole description.

If you have any alternative, plz tell me.

ytakashina committed:
Added description for the return value of GroupBy.{any, all}

pep8speaks commented:
Hello @ytakashina! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/groupby/groupby.py`](https://github.com/pandas-dev/pandas/blob/4de58962265a0e2d99ae8bb087238a649e2a3526/pandas/core/groupby/groupby.py):

> [Line 1334:89](https://github.com/pandas-dev/pandas/blob/4de58962265a0e2d99ae8bb087238a649e2a3526/pandas/core/groupby/groupby.py#L1334): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (123 > 88 characters)
> [Line 1352:89](https://github.com/pandas-dev/pandas/blob/4de58962265a0e2d99ae8bb087238a649e2a3526/pandas/core/groupby/groupby.py#L1352): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (125 > 88 characters)

ytakashina committed:
Fixed violation of PEP8

ytakashina commented on pandas/core/groupby/groupby.py:
```diff
@@ -1330,7 +1330,9 @@ def any(self, skipna: bool = True):
         Returns
         -------
-        bool
+        Series or DataFrame
+            DataFrame or Series of boolean values, which is True if any element is True
+            within each group, False otherwise.
```
I've pushed the third draft for now. But I'm willing to modify these descriptions if requested.

ytakashina committed:
Fixed violation of PEP8

rhshadrach commented on pandas/core/groupby/groupby.py:
```diff
@@ -1330,7 +1330,9 @@ def any(self, skipna: bool = True):
         -------
-        bool
+        Series or DataFrame
+            DataFrame or Series of boolean values, which is True if any element is
```
What do you think about: "..., where a value is True if any element is True within its respective group..."

ytakashina replied:
Sounds more precise than the current one. Thank you!

rhshadrach commented on pandas/core/groupby/groupby.py:
```diff
@@ -1330,7 +1330,9 @@ def any(self, skipna: bool = True):
-        bool
+        Series or DataFrame
+            DataFrame or Series of boolean values, which is True if any element is
+            True within each group, False otherwise.
```
Should this truthful instead of True?

ytakashina replied:
It looks `True` is commonly used in pandas documentation.

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.any.html
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.all.html

`True` (capitalized) is not just a noun, but it represents the boolean literal in Python. So I believe `True` is appropriate in this case.

rhshadrach replied:
Thanks for checking this - agreed.

rhshadrach commented on pandas/core/groupby/groupby.py:
```diff
@@ -1347,7 +1349,9 @@ def all(self, skipna: bool = True):
-        bool
+        Series or DataFrame
+            DataFrame or Series of boolean values, which is True if all elements are
+            True within each group, False otherwise.
```
Similar to the above.

ytakashina committed:
Modified GroupBy.{any, all} docstring

ytakashina commented:
Edited the description for the return value of GroupBy.{any, all} after receiving @rhshadrach's suggestion.

jbrockmendel commented:
Seems like a clear improvement over the status quo.  @rhshadrach are you happy here?

rhshadrach reviewed (approved):
Looks great - thanks @ytakashina

rhshadrach commented:
@ytakashina Can you merge master? I think it will fix CI.

ytakashina committed:
Merge remote-tracking branch 'upstream/master' into ytakashina-patch-type-hinting-groupby

ytakashina committed:
Merge remote-tracking branch 'upstream/master' into ytakashina-patch-type-hinting-groupby

ytakashina committed:
Merge remote-tracking branch 'upstream/master' into ytakashina-patch-type-hinting-groupby

ytakashina commented:
@rhshadrach CI Fixed!

rhshadrach commented:
Thanks @ytakashina!
